go/tasks/plugins/webapi/grpc/plugin.go

```diff
@@ -0,0 +1,189 @@
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &ResourceMetaWrapper{
+		OutputPrefix: outputPrefix,
+		JobID:        res.GetJobId(),
+		Token:        "",
+		TaskType:     taskTemplate.Type,
+	}, &ResourceWrapper{State: service.State_RUNNING}, nil
+}
+
+func (p Plugin) Get(ctx context.Context, taskCtx webapi.GetContext) (latest webapi.Resource, err error) {
+	metadata := taskCtx.ResourceMeta().(*ResourceMetaWrapper)
+
+	client, conn, err := p.getClient(getFinalEndpoint(metadata.TaskType, p.cfg.DefaultGrpcEndpoint, p.cfg.EndpointForTaskTypes))
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect external plugin service")
+	}
+	if conn != nil {
+		defer conn.Close()
+	}
+
+	res, err := client.GetTask(ctx, &service.TaskGetRequest{TaskType: metadata.TaskType, JobId: metadata.JobID})
+	if err != nil {
+		return nil, err
+	}
+
+	return &ResourceWrapper{
+		State:   res.State,
+		Outputs: res.Outputs,
+	}, nil
+}
+
+func (p Plugin) Delete(ctx context.Context, taskCtx webapi.DeleteContext) error {
+	if taskCtx.ResourceMeta() == nil {
+		return nil
+	}
+	metadata := taskCtx.ResourceMeta().(ResourceMetaWrapper)
+
+	client, conn, err := p.getClient(getFinalEndpoint(metadata.TaskType, p.cfg.DefaultGrpcEndpoint, p.cfg.EndpointForTaskTypes))
+	if err != nil {
+		return fmt.Errorf("failed to connect external plugin service")
+	}
+	if conn != nil {
+		defer conn.Close()
+	}
+	_, err = client.DeleteTask(ctx, &service.TaskDeleteRequest{TaskType: metadata.TaskType, JobId: metadata.JobID})
+	return err
+}
+
+func (p Plugin) Status(ctx context.Context, taskCtx webapi.StatusContext) (phase core.PhaseInfo, err error) {
+	resource := taskCtx.Resource().(*ResourceWrapper)
+	taskInfo := &core.TaskInfo{}
+
+	switch resource.State {
+	case service.State_RUNNING:
+		return core.PhaseInfoRunning(pluginsCore.DefaultPhaseVersion, taskInfo), nil
+	case service.State_PERMANENT_FAILURE:
+		return core.PhaseInfoFailure(pluginErrors.TaskFailedWithError, "failed to run the job", taskInfo), nil
+	case service.State_RETRYABLE_FAILURE:
+		return core.PhaseInfoRetryableFailure(pluginErrors.TaskFailedWithError, "failed to run the job", taskInfo), nil
+	case service.State_SUCCEEDED:
+		if resource.Outputs != nil {
+			err := taskCtx.OutputWriter().Put(ctx, ioutils.NewInMemoryOutputReader(resource.Outputs, nil, nil))
+			if err != nil {
+				return core.PhaseInfoUndefined, err
+			}
+		}
+		return core.PhaseInfoSuccess(taskInfo), nil
+	}
+	return core.PhaseInfoUndefined, pluginErrors.Errorf(pluginsCore.SystemErrorCode, "unknown execution phase [%v].", resource.State)
+}
+
+func getFinalEndpoint(taskType, defaultEndpoint string, endpointForTaskTypes map[string]string) string {
+	if t, exists := endpointForTaskTypes[taskType]; exists {
+		return t
+	}
+
+	return defaultEndpoint
+}
+
+func getClientFunc(endpoint string) (service.ExternalPluginServiceClient, *grpc.ClientConn, error) {
```

Comment:
So the endpoint is determined by the taskType with a fallback to a default endpoint, but we should still at least create a cache of gRPC connections (ie `map[endpoint]grpcConn`) so that we don't incur network overhead on each call. Initializing a new TCP connection for each call will make scaling very inefficient.